### PR TITLE
feat: create new package for cookie utilities

### PR DESCRIPTION
**Pull request by saikumarrs**
## PR Description

I've extracted a new package `analytics-js-cookie` to host the utilities related to RudderStack JS SDK cookies.

To avoid, repetitive code, naturally, code had to be shuffled around from other projects.

## Linear task (optional)

https://linear.app/rudderstack/issue/SDK-2036/create-a-separate-npm-package-for-rs-cookie-utilities-js-sdk

## Cross Browser Tests

Please confirm you have tested for the following browsers:

- [x] Chrome
- [ ] Firefox
- [x] IE11

## Sanity Suite

- [x] All sanity suite test cases pass locally

## Security

- [x] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

## Summary by CodeRabbit

- **New Features**
  - Introduced `getDecryptedCookie` function for decrypting and retrieving RudderStack JavaScript SDK cookie values.
  - Added constants `anonymousUserIdKey` and `userTraitsKey` for cookie retrieval.

- **Documentation**
  - Added `README.md` for `@rudderstack/analytics-js-cookies`.
  - Introduced `CHANGELOG.md` for tracking changes.

- **Miscellaneous**
  - Updated `LICENSE` to include MIT License terms.
  - Introduced size limit configuration and Jest configuration for testing.

- **Tests**
  - Added test cases for cookie encryption and decryption functions.

### CODE REVIEW

File: packages/analytics-js-cookies/rollup.config.mjs [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+      ],
+      output: {
+        file: `${outDir}/index.d.ts`,
+        format: 'es',
+      },
+    }
+  ];
+}
+
+export default buildEntries();
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply packages/analytics-js-cookies/rollup.config.mjs:173]: I'll address this later.

File: packages/analytics-js/__tests__/components/eventManager/utilities.test.ts [hunk 2]
```diff
@@ -218,7 +218,7 @@ describe('Event Manager - Utilities', () => {
         title: pageProperties.title,
         url: pageProperties.url,
         referring_domain: pageProperties.referring_domain,
-        tab_url: 'http://www.test-host.com/',
+        tab_url: 'https://www.test-host.com/',
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply packages/analytics-js/__tests__/components/eventManager/utilities.test.ts:221]: It's fine.

File: packages/analytics-js/__tests__/components/configManager/ConfigManager.test.ts [hunk 3]
```diff
@@ -259,7 +259,7 @@ describe('ConfigManager', () => {
 
     configManagerInstance.init();
 
-    expect(state.serverCookies.dataServiceUrl.value).toBe('http://test-host.com/rsaRequest');
+    expect(state.serverCookies.dataServiceUrl.value).toBe('https://test-host.com/rsaRequest');
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply packages/analytics-js/__tests__/components/configManager/ConfigManager.test.ts:262]: It is in the Jest configuration preset file.

File: packages/analytics-js-cookies/src/index.ts [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+  userIdKey,
+  userTraitsKey,
+  anonymousUserIdKey,
+  groupIdKey,
+  groupTraitsKey,
+  initialReferrerKey,
+  initialReferringDomainKey,
+  sessionInfoKey,
+  authTokenKey,
+} from './constants/cookies';
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply packages/analytics-js-cookies/src/index.ts]: Done.

File: packages/analytics-js-cookies/CHANGELOG.md [hunk 5]
```diff
@@ -0,0 +1,16 @@
+# Changelog
+
+This file was generated using [@jscutlery/semver](https://github.com/jscutlery/semver).
+
+## [3.5.0](https://github.com/rudderlabs/rudder-sdk-js/compare/@rudderstack/analytics-js-common@3.4.0...@rudderstack/analytics-js-common@3.5.0) (2024-06-21)
+
+
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply packages/analytics-js-cookies/CHANGELOG.md]: Done.

File: packages/analytics-js-cookies/src/component-cookie/index.ts [hunk 6]
```diff
@@ -2,7 +2,7 @@ import { isNull } from '@rudderstack/analytics-js-common/utilities/checks';
 import type { Nullable } from '@rudderstack/analytics-js-common/types/Nullable';
 import type { CookieOptions } from '@rudderstack/analytics-js-common/types/Storage';
 import type { ILogger } from '@rudderstack/analytics-js-common/types/Logger';
-import { COOKIE_DATA_ENCODING_ERROR } from '../../../constants/logMessages';
+import { COOKIE_DATA_ENCODING_ERROR } from '../constants/logMessages';
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply packages/analytics-js-cookies/src/component-cookie/index.ts]: Not required.

File: packages/analytics-js-cookies/src/component-cookie/index.ts [hunk 7]
```diff
@@ -62,7 +62,7 @@ const set = (
   optionsConfig?: CookieOptions,
   logger?: ILogger,
 ) => {
-  const options: CookieOptions = { ...optionsConfig } || {};
+  const options: CookieOptions = { ...(optionsConfig || {}) };
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply packages/analytics-js-cookies/src/component-cookie/index.ts]: Not required.

File: packages/analytics-js-cookies/__tests__/cookieUtilities.test.ts [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+      document.cookie = 'rl_user_id=RS_ENC_v3_InRlc3QtdXNlcl9pZCI=';
+      document.cookie = 'rl_trait=RS_ENC_v3_ewogICJ0ZXN0LWtleSI6ICJ0ZXN0LXZhbHVlIgp9';
+      document.cookie = 'rl_anonymous_id=RS_ENC_v3_InRlc3QtZGF0YSI=';
+      document.cookie = 'rl_group_id=RS_ENC_v3_InRlc3QtZ3JvdXBfaWQi';
+      document.cookie = 'rl_group_trait=RS_ENC_v3_ewogICJ0ZXN0LWtleSI6ICJ0ZXN0LXZhbHVlIgp9';
+      document.cookie = 'rl_page_init_referrer=RS_ENC_v3_InRlc3QtcGFnZV9pbml0X3JlZmVycmVyIg==';
+      document.cookie =
+        'rl_page_init_referring_domain=RS_ENC_v3_InRlc3QtcGFnZV9pbml0X3JlZmVycmluZ19kb21haW4i';
+      document.cookie = 'rl_session=RS_ENC_v3_ewogICJ0ZXN0LWtleSI6ICJ0ZXN0LXZhbHVlIgp9';
+      document.cookie = 'rl_auth_token=RS_ENC_v3_InRlc3QtdG9rZW4i';
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply packages/analytics-js-cookies/__tests__/cookieUtilities.test.ts]: That's not an API key.

File: packages/analytics-js-cookies/__tests__/cookieUtilities.test.ts [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+  userIdKey,
+  userTraitsKey,
+} from '../src';
+import { encrypt, decrypt, getDecryptedCookie } from '../src/cookiesUtilities';
+
+describe('Cookie Utilities', () => {
+  describe('encrypt', () => {
+    it('should encrypt the value', () => {
+      expect(encrypt('test-data')).toBe('RS_ENC_v3_dGVzdC1kYXRh');
+    });
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply packages/analytics-js-cookies/__tests__/cookieUtilities.test.ts]: We don't need robustness for this use case.

File: packages/analytics-js-cookies/README.md [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+
+- If the cookie is not present.
+- If the cookie is not properly encrypted.
+  - It only decrypts the cookies that are created by the RudderStack JavaScript SDK encryption version "v3".
+- If the decrypted cookie value is not a valid JSON string.
+- If the provided cookie name is not a valid RudderStack JavaScript SDK cookie name.
+
+> If unencrypted, the cookie value will be returned as is.
+
+> Any errors during decryption are swallowed by the function, returning `null`.
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply packages/analytics-js-cookies/README.md:44]: It's fine.

**Review by MoumitaM — APPROVED** [review]

**Pull request merged**